### PR TITLE
Leverage inventory control stream for instance version metrics

### PR DESCRIPTION
**Pull request by rosstimothy**
The version metrics relied on loading all resources in the cache on each auth server to tally the counts. Unmarshalling items from the cache can be quite expensive, especially when done every 15s for all resources in a cluster. While auth servers are typically robust machines the extra computation for calculating a single metric is a bit costly and will grow with the cluster size.

Now that the inventory control stream provides a simple and efficient manner to iterate over all connected instances we can leverage it instead of the cache to calculate the version counts. This is a slight breaking change however, as an auth server will now only be reporting the versions of instances connected directly to itself, and not versions for all instances within the cluster. This can easily be worked around by aggregating the metric from all auth servers for a cluster.

Closes #16429

### CODE REVIEW

**Review by hugoShaka — COMMENTED** [review]
I think we should change the metric Help string to specify this is now per-agent and should be aggregated: https://github.com/gravitational/teleport/blob/a7b72f307c0e409aa7203b14904aa44a06988fa2/lib/auth/auth.go#L348-L349

Regarding the breaking change, I wonder if it would be better to rename the metric so no one tries to use it with a `topk(1, metric)`, this would break for more users, but at least users that already relied on this metric would not get surprised with dashboards showing only half of the agents.

**Review by hugoShaka — APPROVED** [review]

File: docs/pages/reference/metrics.mdx [hunk 1]
```diff
@@ -132,7 +132,7 @@ Teleport Cloud does not expose monitoring endpoints for the Auth Service and Pro
 | `s3_requests_seconds` | histogram | Amazon S3 | Request latency for the S3 API. |
 | `teleport_audit_emit_events` | counter | Teleport Audit Log | Number of audit events emitted. |
 | `teleport_connected_resources` | gauge | Teleport Auth | Number and type of resources connected via keepalives. |
-| `teleport_registered_servers` | gauge | Teleport Auth | The number of Teleport servers (a server consists of one or more Teleport services) that have connected to the Teleport cluster, including the Teleport version. After disconnecting, a Teleport server has a TTL of 10 minutes, so this value will include servers that have recently disconnected but have not reached their TTL. |
+| `teleport_registered_servers` | gauge | Teleport Auth | The number of Teleport services that are connected to an auth server by version. |
```
Comment (jakule) [inline docs/pages/reference/metrics.mdx:135]: Grouped by version?

**Review by jakule — APPROVED** [review]

File: docs/pages/reference/metrics.mdx [hunk 2]
```diff
@@ -132,7 +132,7 @@ Teleport Cloud does not expose monitoring endpoints for the Auth Service and Pro
 | `s3_requests_seconds` | histogram | Amazon S3 | Request latency for the S3 API. |
 | `teleport_audit_emit_events` | counter | Teleport Audit Log | Number of audit events emitted. |
 | `teleport_connected_resources` | gauge | Teleport Auth | Number and type of resources connected via keepalives. |
-| `teleport_registered_servers` | gauge | Teleport Auth | The number of Teleport servers (a server consists of one or more Teleport services) that have connected to the Teleport cluster, including the Teleport version. After disconnecting, a Teleport server has a TTL of 10 minutes, so this value will include servers that have recently disconnected but have not reached their TTL. |
+| `teleport_registered_servers` | gauge | Teleport Auth | The number of Teleport services that are connected to an auth server grouped by version. |
```
Comment (ptgott) [inline docs/pages/reference/metrics.mdx:135]:
```suggestion
| `teleport_registered_servers` | gauge | Teleport Auth | The number of Teleport services that are connected to an Auth Service instance, grouped by version. |
```
So we use terms consistently throughout the docs

**Review by ptgott — APPROVED** [review]
Approved with a suggestion. Disabled auto merge so you can apply it.

File: lib/auth/auth.go [hunk 3]
```diff
@@ -338,16 +338,15 @@ var (
...
 		},
 	)
 
 	registeredAgents = prometheus.NewGaugeVec(
 		prometheus.GaugeOpts{
 			Namespace: teleport.MetricNamespace,
 			Name:      teleport.MetricRegisteredServers,
-			Help: "The number of Teleport servers (a server consists of one or more Teleport services) that have connected to the Teleport cluster, including the Teleport version. " +
-				"After disconnecting, a Teleport server has a TTL of 10 minutes, so this value will include servers that have recently disconnected but have not reached their TTL.",
+			Help:      "The number of Teleport services that are connected to an auth server by version.",
```
Comment (alexfornuto) [inline lib/auth/auth.go:349]: Should @ptgott's update also have been applied here?

Comment (ptgott) [reply lib/auth/auth.go:349]: Created an issue: https://github.com/gravitational/teleport/issues/16734

**Pull request merged**